### PR TITLE
Refactor setup.py external module selection for Cython

### LINKED ISSUE #516: Refactor setup.py external module selection for Cython
Reduce complexity and duplication of the external module selection for Cython in setup.py.

At the time of this issue, this cruft was located in: 
https://github.com/kgriffs/falcon/blob/routing/setup.py#L35

### PR DESCRIPTION
**Pull request by maxking**
Fixes #516 
I am not very sure, but hopefully this is what the issue was about?

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -44,21 +44,12 @@ def list_modules(dirname):
...
-
-    ext_modules += [
-        Extension('falcon.routing.' + ext,
-                  [path.join('falcon', 'routing', ext + '.py')])
+    dirs = ['falcon', 'falcon.util', 'falcon.routing']
+    ext_modules = []
 
-        for ext in list_modules(path.join(MYDIR, 'falcon', 'routing'))]
+    for each in dirs:
+        ext_modules += [Extension(each + ext, [path.join(each, ext + '.py')])
```
Comment (kgriffs) [inline setup.py]: `each` will need to be split on `.` and then re-joined into a path, e.g.:

```py
path.join(*each.split('.'))
```

Also, the module name is missing a `.` between `each` and `ext`.

File: setup.py [hunk 2]
```diff
@@ -44,21 +44,12 @@ def list_modules(dirname):
...
+                            for ext in list_modules(path.join(MYDIR, each))]
```
Comment (kgriffs) [inline setup.py]: Similar to the line above, `each` here is a dotted module name, and so would need to be split first.

**Comment by kgriffs** [comment]
Hi @maxking, just following up to see where we are on this PR

**Comment by maxking** [comment]
@kgriffs Oh, I missed your comments on this one. I will update this over the weekend.

File: setup.py [hunk 3]
```diff
@@ -52,21 +52,13 @@ def list_modules(dirname):
...
-    ext_modules += [
-        Extension('falcon.routing.' + ext,
-                  [path.join('falcon', 'routing', ext + '.py')])
+    dirs = ['falcon', 'falcon.util', 'falcon.routing']
+    ext_modules = []
 
-        for ext in list_modules(path.join(MYDIR, 'falcon', 'routing'))]
+    for each in dirs:
+        ext_modules += [Extension(each + '.' + ext,
+                        [path.join(*each.split('.'), ext + '.py')])
```
Comment (kgriffs) [inline setup.py]: This looks good, just needs a little modification. Currently, running this fails with `SyntaxError: only named arguments may follow *expression`. One way you might work around this is by using the extend operator (`+`), e.g.:

```py
path.join(*(each.split('.') + [ext + '.py'])) 
```

**Comment by kgriffs** [comment]
Hi @maxking, just checking in to see if you had a chance to review my last comment?

**Comment by maxking** [comment]
Hi! @kgriffs, I am afraid, I did not. Have been a little busy. I will update this over the weekend and let you know about it.

**Comment by kgriffs** [comment]
Thanks! Merged as part of #741

**Pull request closed without merging**